File: 卡码/代码随想录/Go/回溯/01-组合/LeetCode_0077.go
//77. 组合
package main

func combine(n int, k int) [][]int {
	path:=make([]int,0)
	res:=make([][]int,0)
	var backtracking func(n,k,index int)
	backtracking=func(n,k,index int){
		if k==len(path) {
			tmp:=make([]int,k)
			copy(tmp,path)
			res=append(res,tmp)
			return
		}
		for i:=index;i<=n;i++ {
			if n+1-i<k-len(path) {
				break
			}
			path=append(path,i)
			backtracking(n,k,i+1)
			path=path[:len(path)-1]
		}
	}
	backtracking(n,k,1)
	return res
}

File: 中等/Java/src/com/kurumi/leetCode/回溯/LeetCode_0077.java
package com.kurumi.leetCode.回溯;

import java.util.ArrayList;
import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-05 下午2:24:24
 * @Description 77. 组合
 */
public class LeetCode_0077 {
    static class Solution {
        private List<List<Integer>> res = new ArrayList<>();

        private void backtracking(int n, int k, int index, List<Integer> path) {
            if (path.size() == k) {
                res.add(new ArrayList<>(path));
                return;
            }
            for (int i = index; n - i + 1 >= k - path.size(); i++) {
                path.add(i);
                backtracking(n, k, i + 1, path);
                path.remove(path.size() - 1);
            }
        }

        public List<List<Integer>> combine(int n, int k) {
            backtracking(n, k, 1, new ArrayList<>());
            return res;
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.combine(4, 2));
    }
}
